src/stan_math_backend/Lower_stmt.ml

```diff
@@ -191,6 +191,10 @@ let rec lower_statement Stmt.Fixed.{pattern; meta} : stmt list =
   location
   @
   match pattern with
+  | Assignment (l, _, e)
+    when Expr.Typed.compare (Stmt.Helpers.expr_of_lvalue l ~meta:e.meta) e = 0
```

Comment:
[Optional] Technically, it's possible for an index expression to have side effects so you should check with `Analysis_and_optimization.Optimize.cannot_remove_expr` but that's a really far-fetched edge case.

Reply:
Oh, it's pretty easy to write a case someone _could_ want to do:

```
generated quantities {
  array[20] real gq;
  for (i in 1:20)
    gq[to_int(uniform_rng(1, 20))] = gq[to_int(uniform_rng(1, 20))];
}
```

I think I'll change it to be more conservative and let the dead code elimination pass in the optimization suite worry about indexed expressions